rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+
+## Internal Proposal
+
+### How and where the pipelines are stored
+
+To keep the actual structure and avoid retro compatibility issues, vector should read from the default folder `/etc/vector/pipelines/*` the difference pipeline's configuration files.
+The pipeline directory should have a flat structure (no sub directories) and only `json`, `toml` and `yaml` files will be supported.
+
+To discuss: make this folder should be configurable through the cli, the same way than for the configuration file, with `--pipelines-dir` or though and environment variable `VECTOR_PIPELINES_DIR`.
+
+### How vector reads the pipelines and what are the limits
+
+Vector will read each of the pipeline configuration files at start time and create an `id` attribute corresponding to the pipeline's filename. A `load-balancer.yml` will have `load-balancer` as `id`. In the pipeline's configuration file, the `id` can be set manually.
+In addition, a `version` attribute will be set containing the hash of the pipeline's configuration file to keep track of the file changes.
+In case several files having the same `id` in that folder (for example `load-balancer.yml` and `load-balancer.json`), vector should error.
+
+Those pipeline's configuration files should only contain transforms or vector will error.
+Each of the defined transforms will be callable by using `pipeline-id.transform-id` outside of the pipeline's configuration file to avoid conflicts.
+A pipeline cannot use a component defined in another pipeline of vector will error.
+
+The structure of the pipeline's configuration file should be as follow
+
+```toml
+id = "load-balancer"
+
+[transforms.first]
+...
+
+[transforms.second]
+...
+```
+
+Now, if we look deeper at the configuration building process, the configuration compiler will require the pipelines in order to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+To do so, we'll need to implement a `PipelineBuilder` that would work the same way as the the `ConfigBuilder` and use it in [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4) in order to load the pipeline's components inside the configuration. When building the `Config`, the [transforms](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L34) from the pipelines will have to be injected only when they are being used.
+
+### What metadatas vector adds to the events going through a pipeline
+
+Each time an event goes through a pipeline, vector will create a field (a tag for metrics) `pipelines.${PIPELINE_ID} = ${PIPELINE_VERSION}` in order to be able to track the path taken by an event.
+
+To do so, here are 2 options.
```

Comment:
I don't care for either of these options. The first 'By dynamically creating a transform each time a pipeline is used' will bottle-neck in remap and the per-event approach taken here retards efforts to move events around in blocks internal to vector, a key performance goal. The second 'By adding the field each time an event goes through a pipeline' makes the transform busier and more solidly per-event, a problem, and adds dynamic work per-event, another problem. 

There's a question I have here that underlay these two concerns: do pipelines ever cross over? That is, do we intend for pipeline 'foo' to ever source or sink into a pipeline 'bar' or will a pipeline 'foo' ever split into separate pipelines 'baz' and 'bing'? It looks like as if we are designing here for merge and split operations but I'm not sure that's actually a requirement. And, if we do design that in I think we're opting into issues around machine locality that we do need to solve for straight out. 

To that last concern, consider being an infra lead at some org in growth stage. You've shipped vector, you have a bunch of pipelines, call them 'alpha', 'beta', 'gamma', 'gamma+alpha'. and they all comfortably fit on one machine. 'gamma+alpha' here means the 'alpha' and 'gamma' pipelines are merged at some point. Now I, the infra lead, know that 'alpha' is going to scale up dramatically so I move it to a new vector deployment. What do I do with 'gamma+alpha'? How does vector deal with that, in this pipeline setup?

Reply:
Unless it's a hard requirement, I'd argue that pipelines should be fully separate in their transform stages. This implies duplicating the source stream into the top of each pipeline and muxing into sinks but these are things we already more or less do. We inject any pipeline tags/labels/whatever at the top of the pipe, which can be done in bulk as needed. Pipeline management from the user perspective then becomes juggling your pipelines to whatever appropriately sized machine without reference to what it may be co-local to.

Reply:
Put another way, if I'm a big enough org that the opening statement of this RFC holds true then I don't have the capability to decide / know which pipelines are co-local to one another and could be split up. Operationally fully separate pipelines are ideal: it opens up the possibility of vector-per-machine-per-pipeline migrations and avoids my vector install slowly becoming a rat king.

Reply:
> That is, do we intend for pipeline 'foo' to ever source or sink into a pipeline 'bar' or will a pipeline 'foo' ever split into separate pipelines 'baz' and 'bing'?

Not at this time. We do plan on allowing the sharing of boilerplate, but I am very concerned about creating a web of pipeline dependencies for management purposes. Constraints are our friend right now and I'd prefer to over constrain to start while we learn.